src/VisualStudio/Core/Def/ProjectSystem/VisualStudioWorkspaceImpl_SourceGenerators.cs

```diff
@@ -0,0 +1,176 @@
+        if (ServiceProvider.GlobalProvider.GetService(typeof(IVsMonitorSelection)) == null)
+            return;
+
+        _isSubscribedToSourceGeneratorImpactingEvents = true;
+
+        // This pattern ensures that we are called whenever the build starts/completes even if it is already in progress.
+        KnownUIContexts.SolutionBuildingContext.WhenActivated(() =>
+        {
+            KnownUIContexts.SolutionBuildingContext.UIContextChanged += (_, e) =>
+            {
+                if (!e.Activated)
+                {
+                    // After a build occurs, transition the solution to a new source generator version.  This will
+                    // ensure that any cached SG documents will be re-generated.
+                    this.EnqueueUpdateSourceGeneratorVersion(projectId: null);
+                }
+            };
+        });
+
+        KnownUIContexts.SolutionExistsAndFullyLoadedContext.WhenActivated(() =>
+        {
+            KnownUIContexts.SolutionExistsAndFullyLoadedContext.UIContextChanged += (_, e) =>
+            {
+                if (e.Activated)
+                {
+                    // After the solution fully loads, transition the solution to a new source generator version.  This
+                    // will ensure that we'll now produce correct SG docs with fully knowledge of all the user's state.
+                    this.EnqueueUpdateSourceGeneratorVersion(projectId: null);
+                }
+            };
+        });
+
+        // Whenever the workspace status changes, go attempt to update generators.
+        var workspaceStatusService = this.Services.GetRequiredService<IWorkspaceStatusService>();
+        workspaceStatusService.StatusChanged += (_, _) => EnqueueUpdateSourceGeneratorVersion(projectId: null);
+
+        // Now kick off at least the initial work to run generators.
+        this.EnqueueUpdateSourceGeneratorVersion(projectId: null);
+    }
+
+    private void EnqueueUpdateSourceGeneratorVersion(ProjectId? projectId)
+        => _updateSourceGeneratorsQueue.AddWork(projectId);
+
+    private async ValueTask ProcessUpdateSourceGeneratorRequestAsync(
+        ImmutableSegmentedList<ProjectId?> projectIds, CancellationToken cancellationToken)
+    {
+        // Only need to do this if we're not in automatic mode.
+        var configuration = this.Services.GetRequiredService<IWorkspaceConfigurationService>().Options;
+        if (configuration.SourceGeneratorExecution is SourceGeneratorExecutionPreference.Automatic)
+            return;
+
+        // Ensure we're fully loaded before rerunning generators.
+        var workspaceStatusService = this.Services.GetRequiredService<IWorkspaceStatusService>();
+        await workspaceStatusService.WaitUntilFullyLoadedAsync(cancellationToken).ConfigureAwait(false);
+
+        var projectIdSet = projectIds.Contains(null) ? null : (ImmutableHashSet<ProjectId>)projectIds.ToImmutableHashSet()!;
```

Comment:
💡 The hash set should not be necessary here. If `projectIds` does not contain `null`, we automatically know that it contains only non-null distinct values because of the way AsyncBatchingWorkQueue coalesces duplicate values.